haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,215 @@
+            else:
+                raise ValueError("question_producer list must contain dicts with keys 'question' and 'document'")
+        else:
+            raise ValueError("Provide either a QuestionGenerator or nonempty list of questions/document pairs")
+
+        if retriever is None:
+            raise ValueError("Retriever must be provided")
+        self.retriever = retriever
+        self.cross_encoder = CrossEncoder(cross_encoder_model_name_or_path)
+        self.total_number_of_questions = total_number_of_questions
+        self.top_k = top_k
+
+    def generate_questions(self, documents: List[Document]) -> List[Dict[str, str]]:
+        """
+        It takes a list of documents and generates a list of question-document pairs
+
+        :param documents: A list of documents to generate questions from
+        :type documents: List[Document]
+        :return: A list of question-document pairs.
+        """
+        question_doc_pairs: List[Dict[str, str]] = []
+        if self.question_document_pairs:
+            question_doc_pairs = self.question_document_pairs
+        else:
+            done = False
+            for idx, doc in enumerate(documents):
+                questions = self.question_generator.generate(text=doc.content)
+                for q in questions:
+                    question_doc_pairs.append({"question": q.strip(), "document": doc.content})
+                    done = len(question_doc_pairs) >= self.total_number_of_questions
+                    if done:
+                        break
+                if done:
+                    break
+        return question_doc_pairs
+
+    def mine_negatives(self, question_doc_pairs: List[Dict[str, str]], batch_size: int = 1) -> List[Dict[str, str]]:
+        """
+        Given a list of question and pos_doc pairs, this function returns a list of question/pos_doc/neg_doc
+        dictionaries
+
+        :param question_doc_pairs: A list of question/pos_doc pairs
+        :type question_doc_pairs: List[Dict[str, str]]
+        :param batch_size: The number of queries to run in a batch, defaults to 1
+        :type batch_size: int (optional)
+        :return: A list of dictionaries, where each dictionary contains the question, positive document,
+                and negative document.
+        """
+        question_pos_doc_neg_doc: List[Dict[str, str]] = []
+
+        for i in range(0, len(question_doc_pairs), batch_size):
+            # question in batches to minimize network latency
+            i_end = min(i + batch_size, len(question_doc_pairs))
+            queries = [e["question"] for e in question_doc_pairs[i:i_end]]
+            pos_docs = [e["document"] for e in question_doc_pairs[i:i_end]]
+
+            # TODO change this to batched queries once they are implemented in retriever
+            # https://github.com/deepset-ai/haystack/issues/1239
+            retriever_results = self.retriever.retrieve(query=queries[0], top_k=self.top_k)
```

Comment:
Yes, we are because I thought the query batching is not available yet. I prepared the code for the eventual switch to batched queries. Are the batched queries supported now?

Reply:
Waiting for #1239